README.md
# orglinter

[![Version npm](https://img.shields.io/npm/v/orglinter.svg)](https://www.npmjs.com/package/orglinter)
[![License](https://img.shields.io/npm/l/orglinter.svg)](https://github.com/godaddy/orglinter/blob/master/LICENSE)
[![npm Downloads](https://img.shields.io/npm/dm/orglinter.svg)](https://npmcharts.com/compare/orglinter?minimal=true)
[![Build Status](https://img.shields.io/github/workflow/status/godaddy/orglinter/Node.js%20CI)](https://github.com/godaddy/orglinter/actions?query=workflow%3A%22Node.js+CI%22)
[![Codecov](https://img.shields.io/codecov/c/github/godaddy/orglinter)](https://codecov.io/gh/godaddy/orglinter)
[![Dependencies](https://img.shields.io/david/godaddy/orglinter.svg)](https://github.com/godaddy/orglinter/blob/master/package.json)

Lint GitHub organizations to ensure proper membership lists and configuration.

## Installation

`orglinter` is designed to be installed with npm or yarn, either globally:

```console
$ npm install -g orglinter
$ orglinter my-org-file.toml
```

Within your package:

```console
$ npm install orglinter
$ npx orglinter my-org-file.toml
```

Or if you would like to make contributions to `orglinter`, using `git` and `npm`
for local development:

```console
$ git clone git@github.com:godaddy/orglinter.git
$ cd orglinter
$ npm install
$ npm start my-org-file.toml
```

## Usage

A stub `.env.dist` file can be found for the convenience of users and developers
at the root of this project. If you copy this file to `.env` and fill in the
blanks, you should be good to go!

Running the script requires a single environment variable, which you can place
into a `.env` file for automatic ingestion: `GITHUB_TOKEN`. This must be a
personal access token generated by an existing organization owner, with the
following permissions:

* `admin:org`
* `read:user`

Once a `GITHUB_TOKEN` is ready, the script can be run by `npm start`.

By default, the tool will look for a file at `./membership.toml`. However you
can also specify a different filename on the command line such as:

```console
$ npm start my-org-file.toml

> orglinter@1.0.0 start
> node ./src/bin/cli.js
# ...
```

At the time of this writing, this script will be run manually after every change
to the membership file. In the future, this will become an automatic execution
in the CI pipeline, and will also be run on a regular basis to ensure that
undocumented memberships do not exist, and that employees who have left the
company are removed from the org.

## Org Specification

All org data is stored in [TOML] files; one file representing one GitHub org. An
example of such a file is provided as `example.toml` at the root of this project.

The relevant sections of this file are:

### [org]

This contains configuration information about the org itself. Things such as the
email address, Twitter username, and website URL. These values will all be
verified against the actual org configuration and will raise an error if they do
not match reality.

Additionally, the `login` key here is what is used to perform the actual org
lookup.

### [members]

This is the membership list for your org. The idea here is to map "internal"
usernames (such as those found in an LDAP directory) to GitHub usernames. So
assuming you have an "internal" username of `adent`, and a GitHub username of
`arthur_dent`, you would add a line to this section like:

```toml
adent = "arthur_dent"
```

You can also tie multiple GitHub usernames to a single "internal" username by
using a list. For example:

```toml
tmcmillan = ["trillian", "trillian_astra"]
```

### [[teams]]

Each `[[teams]]` section is used, for now, primarily as a means of managing the
admins of an org. In the future, this will manage actual GitHub teams. These are
written as an [array of tables]. So one team would look like:

```toml
[[teams]]
name = "Heart of Gold Crew"
default_org_role = "ADMIN"
privacy = "SECRET"
members = ["mandroid", "tmcmillan", "zbeeblebrox"]
```

Note that the `default_org_role = "ADMIN"` is what marks the members of this
team as org admins. And the names used in the `members` list are the "internal"
usernames, rather than the GitHub usernames, of the individual members. This
does have the side effect that, if an internal username is connected to multiple
GitHub usernames, as demonstrated above, all of those GitHub usernames will be
made admins of the org.

## Development

### Running Tests

This project uses [mocha], [assume], and [sinon] for its tests. It also uses
[eslint] for ensuring code styling. Luckily, the execution of these tools is
nicely automated for you. To run all of them, all you need to do is:

```console
$ npm test
```

This will run all of the tests for you, followed by the linter.

## Credits

Originally developed internally for making easier work of the management of
[GoDaddy]'s GitHub orgs. Primary authors at that time include:

* [Joey Wilhelm]
* [Rick Markins]

[array of tables]: https://toml.io/en/v1.0.0#array-of-tables
[assume]: http://assume.fail/
[GoDaddy]: https://www.godaddy.com/
[Joey Wilhelm]: https://github.com/tarkatronic/
[mocha]: https://mochajs.org/
[Rick Markins]: https://github.com/rxmarbles/
[sinon]: https://sinonjs.org/
[TOML]: https://toml.io/


CHANGELOG.md
# `orglinter`

- Fixed #16 - New users were causing an error in findPromotions
- Fixed #20 - Added full JSDoc coverage across the code
- Fixed #23 - Added auditing of application installations

## 0.1.0

- Initial MVP release


test/loaders.spec.js
'use strict';

const assume = require('assume');
const fs = require('fs').promises;
const loaders = require('../src/lib/loaders');
const nock = require('nock');
const path = require('path');
const sinon = require('sinon');


describe('Loaders', function () {

  before(function () {
    sinon.stub(console, 'log');
  });

  after(function () {
    sinon.restore();
  });

  describe('retrieveOrgApplications', function () {
    let scope;

    before(async function () {
      const response = JSON.parse(await fs.readFile(path.resolve(__dirname, 'responses/installations.json')));
      scope = nock('https://api.github.com/').get('/orgs/foo/installations').reply(200, response).persist();
    });

    after(function () {
      nock.cleanAll();
    });

    it('calls the proper REST endpoint', async function () {
      await loaders.retrieveOrgApplications('foo', '12345');
      scope.done();
    });

    it('returns only the relevant information', async function () {
      const apps = await loaders.retrieveOrgApplications('foo', '12345');

      assume(apps).eqls([
        {
          appId: 12,
          appSlug: 'heart-of-gold',
          repositorySelection: 'all',
          permissions: {
            foo: 'read',
            bar: 'write',
            baz: 'read'
          },
          events: ['push', 'release']
        },
        {
          appId: 13,
          appSlug: 'infinite-improbability-drive',
          repositorySelection: 'selected',
          permissions: {
            foo: 'write',
            bar: 'read',
            baz: 'write'
          },
          events: ['issue', 'discussion', 'pull_request']
        }
      ]);
    });
  });
});


test/checkers.spec.js
/* eslint-disable no-console */
'use strict';

const assume = require('assume');
const checkers = require('../src/lib/checkers');
const sinon = require('sinon');

describe('Checkers', function () {

  before(function () {
    sinon.stub(console, 'log');
  });

  after(function () {
    sinon.restore();
  });

  describe('findUndocumentedMembers', function () {
    it('finds members not listed in config', function () {
      const result = checkers.findUndocumentedMembers(
        ['zaphod', 'trillian', 'marvin'],
        ['zaphod', 'trillian', 'marvin', 'ford']
      );
      assume(result).eqls(['ford']);
    });
  });

  describe('findNewMembers', function () {
    it('finds members not yet in the org', function () {
      const result = checkers.findNewMembers(
        ['zaphod', 'trillian', 'marvin', 'ford'],
        ['zaphod', 'trillian', 'marvin']
      );
      assume(result).eqls(['ford']);
    });
  });

  describe('findDemotions', function () {
    it('finds people who should not be admins', function () {
      const result = checkers.findDemotions(
        { zaphod: 'ADMIN', trillian: 'ADMIN', marvin: 'MEMBER', ford: 'MEMBER' },
        { zaphod: { role: 'ADMIN' }, trillian: { role: 'ADMIN' }, marvin: { role: 'MEMBER' }, ford: { role: 'ADMIN' } }
      );
      assume(result).eqls(['ford']);
    });
  });

  describe('findPromotions', function () {
    it('finds people who should be admins', function () {
      const result = checkers.findPromotions(
        { zaphod: 'ADMIN', trillian: 'ADMIN', marvin: 'MEMBER', ford: 'MEMBER' },
        { zaphod: { role: 'MEMBER' }, trillian: { role: 'ADMIN' }, marvin: { role: 'MEMBER' }, ford: { role: 'MEMBER' } }
      );
      assume(result).eqls(['zaphod']);
    });

    it('ignores users not in retrieved list', function () {
      const result = checkers.findPromotions(
        { zaphod: 'ADMIN', trillian: 'ADMIN', marvin: 'MEMBER', ford: 'MEMBER' },
        { zaphod: { role: 'MEMBER' }, marvin: { role: 'MEMBER' }, ford: { role: 'MEMBER' } }
      );
      assume(result).eqls(['zaphod']);
    });
  });

  describe('validateTwoFactor', function () {
    it('finds people somehow not using 2fa', function () {
      const result = checkers.validateTwoFactor(
        {
          zaphod: { twoFactorAuth: true },
          trillian: { twoFactorAuth: true },
          marvin: { twoFactorAuth: true },
          ford: { twoFactorAuth: false }
        }
      );
      assume(result).eqls(['ford']);
    });
  });

});


src/lib/fixers.js
/* eslint-disable no-console */
'use strict';

const { request } = require('@octokit/request');

/**
 * Invite a user to join an org
 *
 * @param {object} invitation - The details of the user invitation to be sent
 * @param {string} invitation.username - The username to be invited
 * @param {string} invitation.organization - The login name of the org where the user will be invited
 * @param {string} invitation.token - A personal access token for interacting with the GitHub API
 * @param {string} [invitation.role=member] - The role to invite the user as; "admin" or "member"
 * @param {boolean} [invitation.dryRun=false] - If true, don't actually execute this operation
 */
async function inviteUser({ username, organization, token, role = 'member', dryRun = false }) {
  console.log(`Setting ${username} up as a(n) ${role} for ${organization}.`);
  if (dryRun) return;
  // Do the thing!
  const result = await request('PUT /orgs/{org}/memberships/{username}', {
    headers: {
      authorization: `token ${token}`
    },
    org: organization,
    username: username,
    role: role
  });
  if (result.status === 200) {
    console.log(`${username} successfully invited to ${organization}!`);
  } else {
    console.log(`ERROR ERROR! ${username} could not be invited to ${organization}!`);
  }
}

/**
 * Remove a user from an org
 *
 * @param {object} removal - The details of the user removal to be initiated
 * @param {string} removal.username - The username to be removed from the org
 * @param {string} removal.organization - The login name of the org to remove the user from
 * @param {string} removal.token - A personal access token for interacting with the GitHub API
 * @param {boolean} removal.dryRun - If true, don't actally execute this operation
 */
async function removeMember({ username, organization, token, dryRun = false }) {
  console.log(`Removing ${username} from ${organization}.`);
  if (dryRun === true) return;
  // Do the thing!
  const result = await request('DELETE /orgs/{org}/members/{username}', {
    headers: {
      authorization: `token ${token}`
    },
    org: organization,
    username: username
  });
  if (result.status === 204) {
    console.log(`${username} successfully removed from ${organization}.`);
  } else {
    console.log(`ERROR ERROR! ${username} could not be removed from ${organization}!`);
  }
}

/**
 * Promote a user from member to admin of an org
 *
 * @param {object} promotion - The details of the user promotion to be sent
 * @param {string} promotion.username - The username to be promoted
 * @param {string} promotion.organization - The login name of the org to promote the user in
 * @param {string} promotion.token - A personal access token for interacting with the GitHub API
 * @param {boolean} promotion.dryRun - If true, don't actally execute this operation
 */
async function promoteMember({ username, organization, token, dryRun = false }) {
  console.log(`Promoting ${username} to admin in ${organization}.`);
  if (dryRun) return;
  // Do the thing!
  await inviteUser({ username, organization, token, role: 'admin' });
}

/**
 * Demote a user from admin to member of an org
 *
 * @param {object} demotion - The details of the user demotion to be sent
 * @param {string} demotion.username - The username to be demoted
 * @param {string} demotion.organization - The login name of the org to demote the user in
 * @param {string} demotion.token - A personal access token for interacting with the GitHub API
 * @param {boolean} demotion.dryRun - If true, don't actally execute this operation
 */
async function demoteMember({ username, organization, token, dryRun = false }) {
  console.log(`Demoting ${username} to member in ${organization}.`);
  if (dryRun) return;
  // Do the thing!
  await inviteUser({ username, organization, token, role: 'member' });
}

module.exports = { inviteUser, removeMember, promoteMember, demoteMember };


src/lib/loaders.js
/* eslint-disable no-console */
'use strict';

const fs = require('fs').promises;
const { graphql } = require('@octokit/graphql');
const { request } = require('@octokit/request');
const TOML = require('@iarna/toml');
const typedefs = require('./typedefs');


/**
 * Retrieve information from the GitHub GraphQL API about the requested org
 *
 * @param {string} orgName - The login name of the org to be retrieved
 * @param {string} token - A personal access token for interacting with the API
 * @returns {typedefs.OrgRecord} - The full details of the retrieved org
 */
async function retrieveOrgInfo(orgName, token) {
  let totalCount = 1, retrieved = 0, after = null, allMembers = [], pendingMembers, organization;
  const perPage = 100;
  while (retrieved < totalCount) {
    console.log(`Retrieving ${perPage} GitHub members...`);
    ({ organization } = await graphql(
      `
      query getAllMembers($organization: String!, $first: Int = 100, $after: String = null)
      {
        organization(login: $organization) {
          email
          isVerified
          login
          name
          requiresTwoFactorAuthentication
          twitterUsername
          websiteUrl
          membersWithRole(first: $first, after: $after) {
            edges {
              role
              hasTwoFactorEnabled
              node {
                login
              }
              cursor
            }
            totalCount
          }
          pendingMembers(first: 50) {
            totalCount
            edges {
              node {
                login
              }
            }
          }
        }
      }
      `,
      {
        organization: orgName,
        first: perPage,
        after: after,
        headers: {
          authorization: `token ${token}`
        }
      }
    ));
    const currMembers = organization.membersWithRole;
    // TODO: Loop through these in case there are more than 50?
    pendingMembers = organization.pendingMembers;
    allMembers = allMembers.concat(currMembers.edges);
    totalCount = currMembers.totalCount;
    retrieved += perPage;
    after = currMembers.edges[currMembers.edges.length - 1].cursor;
  }
  const result = {
    members: {
      ...allMembers.reduce((members, member) => {
        members[member.node.login] = {
          role: member.role,
          twoFactorAuth: member.hasTwoFactorEnabled
        };
        return members;
      }, {}),
      ...pendingMembers.edges.reduce((members, member) => {
        members[member.node.login] = {
          role: 'PENDING',
          twoFactorAuth: null
        };
        return members;
      }, {})
    },
    email: organization.email,
    isVerified: organization.isVerified,
    login: organization.login,
    name: organization.name,
    // eslint-disable-next-line id-length
    requiresTwoFactorAuthentication: organization.requiresTwoFactorAuthentication,
    twitterUsername: organization.twitterUsername,
    websiteUrl: organization.websiteUrl,
    applications: await retrieveOrgApplications(orgName, token)
  };
  console.log(`${Object.keys(result.members).length} total members retrieved.`);
  return result;
}

/**
 * Retrieve a list of all applications installed to the requested org
 *
 * @param {string} orgName - The login name of the org to retrieve applications for
 * @param {string} token - A personal access token for interacting with the API
 * @returns {typedefs.AppSet} - A list of all applications installed on the organization
 */
async function retrieveOrgApplications(orgName, token) {
  const response = await request('GET /orgs/{org}/installations', {
    headers: { authorization: `token ${token}` },
    org: orgName
  });
  const result = response.data.installations.reduce((apps, app) => {
    apps.push({
      appId: app.app_id,
      appSlug: app.app_slug,
      repositorySelection: app.repository_selection,
      permissions: app.permissions,
      events: app.events
    });
    return apps;
  }, []);
  console.log(`Loaded ${result.length} application installations for ${orgName}.`);
  return result;
}

/**
 * Load an org's expected configuration from a TOML config file
 *
 * @param {string} fileName - The full path to the config file
 * @returns {typedefs.ExpectedOrgConfig} - The full expected configuration of the org
 */
async function loadMembershipConfig(fileName) {
  const config = await TOML.parse.async(await fs.readFile(fileName));
  const allAdmins = new Set(
    config.teams.filter(
      team => team.default_org_role === 'ADMIN'
    ).reduce(
      (members, team) => { return members.concat(team.members); },
      []
    )
  );

  // Construct a new object consisting of github usernames and membership roles
  config['github-members'] = Object.keys(config.members).reduce((members, key) => {
    const isAdmin = allAdmins.has(key);
    if (Array.isArray(config.members[key])) {
      config.members[key].forEach((member) => { members[member] = isAdmin ? 'ADMIN' : 'MEMBER'; });
    } else {
      members[config.members[key]] = isAdmin ? 'ADMIN' : 'MEMBER';
    }
    return members;
  }, {});
  // TODO: Validate usernames against LDAP
  console.log(`Loaded config for ${config.org.login}. ${Object.keys(config['github-members']).length} memberships expected.`);
  return config;
}

module.exports = { retrieveOrgInfo, retrieveOrgApplications, loadMembershipConfig };


src/lib/checkers.js
/* eslint-disable no-console */
'use strict';

const typedefs = require('./typedefs');

/**
 * Find users who currently belong to the organization, but not who are not
 * expected by config.
 *
 * @param {string[]} configured - An array of usernames expected to exist in the org
 * @param {string[]} retrieved - An array of usernames actually belong to the org currently
 * @returns {string[]} An array of usernames which are not expected to belong to the org
 */
function findUndocumentedMembers(configured, retrieved) {
  const undocumented = retrieved.filter(member => !configured.includes(member));
  console.log(`${undocumented.length} undocumented memberships found: `, undocumented);
  return undocumented;
}

/**
 * Find users who are expected by config, but who do not (yet) belong to the
 * organization.
 *
 * @param {string[]} configured - An array of usernames expected to exist in the org
 * @param {string[]} retrieved - An array of usernames actually belong to the org currently
 * @returns {string[]} An array of usernames which are expected to belong to the org
 */
function findNewMembers(configured, retrieved) {
  const newMembers = configured.filter(member => !retrieved.includes(member));
  console.log(`${newMembers.length} new memberships requested: `, newMembers);
  return newMembers;
}

/**
 * Find users who are currently admins of the org, but should not be, according
 * to config.
 *
 * @param {object.<string, string>} configured - Configured usernames and their associated roles
 * @param {typedefs.MemberSet} retrieved - User list retrieved from GitHub
 * @returns {string[]} An array of usernames who are admins, but should not be
 */
function findDemotions(configured, retrieved) {
  const demotions = Object.keys(retrieved).filter(
    (member) => retrieved[member].role === 'ADMIN' && configured[member] === 'MEMBER'
  );
  console.log(`${demotions.length} demotions found: `, demotions);
  return demotions;
}

/**
 * Find users who are currently regular members of the org, but are configured as admins
 *
 * @param {object.<string, string>} configured - Configured usernames and their associated roles
 * @param {typedefs.MemberSet} retrieved - User list retrieved from GitHub
 * @returns {string[]} An array of usernames who are not admins, but should be
 */
function findPromotions(configured, retrieved) {
  const promotions = Object.keys(configured).filter(
    (member) => configured[member] === 'ADMIN' && retrieved[member] && retrieved[member].role === 'MEMBER'
  );
  console.log(`${promotions.length} promotions found: `, promotions);
  return promotions;
}

/**
 * Find org members who do not have two-factor authentication enabled
 *
 * @param {typedefs.MemberSet} members - User list retrieved from GitHub
 * @returns {string[]} An array of usernames who are 2FA violators.
 */
function validateTwoFactor(members) {
  const violators = Object.keys(members).filter((member) => members[member].twoFactorAuth === false);
  console.log(`${violators.length} two factor auth violators found: `, violators);
  return violators;
}

/**
 * Ensure that the org's current settings match those expected by the config
 *
 * @param {object.<string, *>} organization - The full collection of org settings collected from GitHub
 * @param {object.<string, *>} config - The full collection of org settings expected by config
 */
function validateOrgSettings(organization, config) {
  let failed = false;
  for (const key of Object.keys(config)) {
    if (organization[key] !== config[key]) {
      console.log(`Configured value does not match retrieved value for ${key}. "${config[key]}" vs "${organization[key]}"`);
      failed = true;
    }
  }
  if (failed === true) {
    console.log(`\nERROR! Misconfigured org settings found! See above for more details.\n`);
  }
}

module.exports = {
  findUndocumentedMembers,
  findNewMembers,
  findDemotions,
  findPromotions,
  validateTwoFactor,
  validateOrgSettings
};


src/lib/typedefs.js
'use strict';

/**
 * A normalized member record retrieved from the GitHub GraphQL API
 *
 * @typedef {object} MemberRecord
 * @property {string} role - The user's role in the org; "ADMIN" or "MEMBER"
 * @property {boolean} hasTwoFactorEnabled - Whether the user has Two Factor Authentication enabled
 */

/**
 * A set of member records retrieved from the GitHub GraphQL API
 *
 * @typedef {object.<string, MemberRecord>} MemberSet
 */

/**
 * A normalized application installation record retrieved from the GitHub REST API
 *
 * @typedef {object} AppRecord
 * @property {number} appId - The global database identifier for the application
 * @property {string} appSlug - A unique slug for identifying the application
 * @property {string} repositorySelection - Whether this is installed for "all" or "selected" repositories
 * @property {object.<string, string>} permissions - The set of permissions granted to this application
 * @property {Array.<string>} events - A list of events that are received by this application
 */

/**
 * A set of application installation records retrieved from the GitHub REST API
 *
 * @typedef {Array.<AppRecord>} AppSet
 */

/**
 * A normalized org record retrieved from the GitHub GraphQL API
 *
 * @typedef {object} OrgRecord
 * @property {string} email - The org's primary email address
 * @property {boolean} isVerified - Whether the org has been verified by GitHub
 * @property {string} login - The org's login name
 * @property {MemberSet} members - The org's current membership
 * @property {string} name - The org's display name
 * @property {boolean} requiresTwoFactorAuthentication - Whether the org requires 2FA for membership
 * @property {string} twitterUsername - The org's twitter username
 * @property {string} websiteURL - The org's primary external website URL
 */

/**
 * The expected configuration of a single GitHub team
 *
 * @typedef {object} ConfigTeam
 * @property {string} name - The display name of the team
 * @property {string} default_org_role - The org role assigned to members of this team; "ADMIN" or "MEMBER"
 * @property {string} privacy - The public visibility of this team; "SECRET" or "VISIBLE"
 * @property {Array.<string>} members - The users who are expected to be members of this team
 */

/**
 * The expected configuration for the org entity itself
 *
 * @typedef {object} ConfigOrg
 * @property {string} email - The org's expected email address
 * @property {boolean} isVerified - Whether the org is expected to be verified by GitHub
 * @property {string} login - The login name for the org; this is used for the API lookup
 * @property {string} name - The org's expected display name
 * @property {boolean} requiresTwoFactorAuthentication - Whether the org is expected to require 2FA for membership
 * @property {string} twitterUsername - The org's expected twitter username
 * @property {string} websiteUrl - The org's expected primary external website URL
 */

/**
 * Structured data representing the expected configuration of a GitHub org
 *
 * @typedef {object} ExpectedOrgConfig
 * @property {ConfigOrg} org - The expected configuration for the org entity itself
 * @property {object.<string, string | Array.<string>>} members - A set of "internal" usernames paired up with GitHub usernames
 * @property {object.<string, string>} github-members - GitHub usernames and their expected roles
 * @property {Array.<ConfigTeam>} teams - The teams expected to exist in the org
 */

module.exports = {};


index.js
/* eslint-disable no-console, no-process-env */
require('dotenv').config();
const fs = require('fs').promises;
const { graphql } = require('@octokit/graphql');
const path = require('path');
const { request } = require('@octokit/request');
const TOML = require('@iarna/toml');

async function retrieveOrgInfo(orgName) {
  let totalCount = 1, retrieved = 0, after = null, allMembers = [], pendingMembers, organization;
  const perPage = 100;
  while (retrieved < totalCount) {
    console.log(`Retrieving ${perPage} GitHub members...`);
    ({ organization } = await graphql(
      `
      query getAllMembers($organization: String!, $first: Int = 100, $after: String = null)
      {
        organization(login: $organization) {
          email
          isVerified
          login
          name
          requiresTwoFactorAuthentication
          twitterUsername
          websiteUrl
          membersWithRole(first: $first, after: $after) {
            edges {
              role
              hasTwoFactorEnabled
              node {
                login
              }
              cursor
            }
            totalCount
          }
          pendingMembers(first: 50) {
            totalCount
            edges {
              node {
                login
              }
            }
          }
        }
      }
      `,
      {
        organization: orgName,
        first: perPage,
        after: after,
        headers: {
          authorization: `token ${process.env.GITHUB_TOKEN}`
        }
      }
    ));
    const currMembers = organization.membersWithRole;
    // TODO: Loop through these in case there are more than 50?
    pendingMembers = organization.pendingMembers;
    allMembers = allMembers.concat(currMembers.edges);
    totalCount = currMembers.totalCount;
    retrieved += perPage;
    after = currMembers.edges[currMembers.edges.length - 1].cursor;
  }
  const result = {
    members: {
      ...allMembers.reduce((members, member) => {
        members[member.node.login] = {
          role: member.role,
          twoFactorAuth: member.hasTwoFactorEnabled
        };
        return members;
      }, {}),
      ...pendingMembers.edges.reduce((members, member) => {
        members[member.node.login] = {
          role: 'PENDING',
          twoFactorAuth: null
        };
        return members;
      }, {})
    },
    email: organization.email,
    isVerified: organization.isVerified,
    login: organization.login,
    name: organization.name,
    // eslint-disable-next-line id-length
    requiresTwoFactorAuthentication: organization.requiresTwoFactorAuthentication,
    twitterUsername: organization.twitterUsername,
    websiteUrl: organization.websiteUrl
  };
  console.log(`${Object.keys(result.members).length} total members retrieved.`);
  return result;
}

async function loadMembershipConfig(fileName) {
  const config = await TOML.parse.async(await fs.readFile(fileName));
  const allAdmins = new Set(
    config.teams.filter(
      team => team.default_org_role === 'ADMIN'
    ).reduce(
      (members, team) => { return members.concat(team.members); },
      []
    )
  );

  // Construct a new object consisting of github usernames and membership roles
  config['github-members'] = Object.keys(config.members).reduce((members, key) => {
    const isAdmin = allAdmins.has(key);
    if (Array.isArray(config.members[key])) {
      config.members[key].forEach((member) => { members[member] = isAdmin ? 'ADMIN' : 'MEMBER'; });
    } else {
      members[config.members[key]] = isAdmin ? 'ADMIN' : 'MEMBER';
    }
    return members;
  }, {});
  // TODO: Validate usernames against LDAP
  console.log(`Loaded config for ${config.org.login}. ${Object.keys(config['github-members']).length} memberships expected.`);
  return config;
}

function findUndocumentedMembers(configured, retrieved) {
  const undocumented = retrieved.filter(member => !configured.includes(member));
  console.log(`${undocumented.length} undocumented memberships found: `, undocumented);
  return undocumented;
}

function findNewMembers(configured, retrieved) {
  const newMembers = configured.filter(member => !retrieved.includes(member));
  console.log(`${newMembers.length} new memberships requested: `, newMembers);
  return newMembers;
}

function findDemotions(configured, retrieved) {
  const demotions = Object.keys(retrieved).filter(
    (member) => retrieved[member].role === 'ADMIN' && configured[member] === 'MEMBER'
  );
  console.log(`${demotions.length} demotions found: `, demotions);
  return demotions;
}

function findPromotions(configured, retrieved) {
  const promotions = Object.keys(configured).filter(
    (member) => configured[member] === 'ADMIN' && retrieved[member].role === 'MEMBER'
  );
  console.log(`${promotions.length} promotions found: `, promotions);
  return promotions;
}

function validateTwoFactor(members) {
  const violators = Object.keys(members).filter((member) => members[member].twoFactorAuth === false);
  console.log(`${violators.length} two factor auth violators found: `, violators);
  return violators;
}

function validateOrgSettings(organization, config) {
  let failed = false;
  for (const key of Object.keys(config)) {
    if (organization[key] !== config[key]) {
      console.log(`Configured value does not match retrieved value for ${key}. "${config[key]}" vs "${organization[key]}"`);
      failed = true;
    }
  }
  if (failed === true) {
    console.log(`\nERROR! Misconfigured org settings found! See above for more details.\n`);
  }
}

async function inviteUser(username, organization, role = 'member', dryRun) {
  console.log(`Setting ${username} up as a(n) ${role} for ${organization}.`);
  if (dryRun) return;
  // Do the thing!
  const result = await request('PUT /orgs/{org}/memberships/{username}', {
    headers: {
      authorization: `token ${process.env.GITHUB_TOKEN}`
    },
    org: organization,
    username: username,
    role: role
  });
  if (result.status === 200) {
    console.log(`${username} successfully invited to ${organization}!`);
  } else {
    console.log(`ERROR ERROR! ${username} could not be invited to ${organization}!`);
  }
}

async function removeMember(username, organization, dryRun) {
  console.log(`Removing ${username} from ${organization}.`);
  if (dryRun === true) return;
  // Do the thing!
  const result = await request('DELETE /orgs/{org}/members/{username}', {
    headers: {
      authorization: `token ${process.env.GITHUB_TOKEN}`
    },
    org: organization,
    username: username
  });
  if (result.status === 204) {
    console.log(`${username} successfully removed from ${organization}.`);
  } else {
    console.log(`ERROR ERROR! ${username} could not be removed from ${organization}!`);
  }
}

async function promoteMember(username, organization, dryRun) {
  console.log(`Promoting ${username} to admin in ${organization}.`);
  if (dryRun) return;
  // Do the thing!
  await inviteUser(username, organization, 'admin');
}

async function demoteMember(username, organization, dryRun) {
  console.log(`Demoting ${username} to member in ${organization}.`);
  if (dryRun) return;
  // Do the thing!
  await inviteUser(username, organization, 'member');
}

(async () => {
  const DRY_RUN = !!process.env.DRY_RUN || false;
  const fileName = path.normalize(process.argv[2] || './membership.toml');
  const pendingActions = [];

  if (DRY_RUN) console.log('NOTE: This is a dry run. Actions will not actually be performed.');
  const config = await loadMembershipConfig(fileName), orgName = config.org.login;
  const organization = await retrieveOrgInfo(orgName);
  validateOrgSettings(organization, config.org);

  // If 2FA is enabled, kick out anybody who somehow doesn't have it
  if (organization.requiresTwoFactorAuthentication) {
    pendingActions.push(
      ...validateTwoFactor(organization.members).map((username) => removeMember(username, orgName, DRY_RUN))
    );
  }

  // Remove members who do not belong
  pendingActions.push(
    ...findUndocumentedMembers(
      Object.keys(config['github-members']),
      Object.keys(organization.members)
    ).map((username) => removeMember(username, orgName, DRY_RUN))
  );
  // Invite members who are not present
  pendingActions.push(
    ...findNewMembers(
      Object.keys(config['github-members']),
      Object.keys(organization.members)
    ).map((username) =>
      inviteUser(username, orgName, config['github-members'][username].toLowerCase(), DRY_RUN)
    )
  );
  // Remove admin status from those who should not have it
  pendingActions.push(
    ...findDemotions(
      config['github-members'],
      organization.members
    ).map((username) => demoteMember(username, orgName, DRY_RUN))
  );
  // Grant admin status to those who should have it
  pendingActions.push(
    ...findPromotions(
      config['github-members'],
      organization.members
    ).map((username) => promoteMember(username, orgName, DRY_RUN))
  );
  await Promise.all(pendingActions);
  console.log(`${pendingActions.length} actions completed successfully!`);
})();
